src/Libraries/Microsoft.Extensions.DataIngestion/Writers/VectorStoreWriter.cs

```diff
@@ -0,0 +1,201 @@
+    /// <param name="options">The options for the vector store writer.</param>
+    /// <exception cref="ArgumentNullException">When <paramref name="vectorStore"/> is null.</exception>
+    /// <exception cref="ArgumentOutOfRangeException">When <paramref name="dimensionCount"/> is less or equal zero.</exception>
+    public VectorStoreWriter(VectorStore vectorStore, int dimensionCount, VectorStoreWriterOptions? options = default)
+    {
+        _vectorStore = Throw.IfNull(vectorStore);
+        _dimensionCount = Throw.IfLessThanOrEqual(dimensionCount, 0);
+        _options = options ?? new VectorStoreWriterOptions();
+
+        // Not all vector store support string as the key type, examples:
+        // Qdrant: https://github.com/microsoft/semantic-kernel/blob/28ea2f4df872e8fd03ef0792ebc9e1989b4be0ee/dotnet/src/VectorData/Qdrant/QdrantCollection.cs#L104
+        // When https://github.com/microsoft/semantic-kernel/issues/13141 gets released,
+        // we need to remove this workaround.
+        _keysAreStrings = vectorStore.GetType().Name != "QdrantVectorStore";
+    }
+
+    /// <summary>
+    /// Gets the underlying <see cref="VectorStoreCollection{TKey,TRecord}"/> used to store the chunks.
+    /// </summary>
+    /// <remarks>
+    /// The collection is initialized when <see cref="WriteAsync(IAsyncEnumerable{IngestionChunk{T}}, CancellationToken)"/> is called for the first time.
+    /// </remarks>
+    /// <exception cref="InvalidOperationException">The collection has not been initialized yet.
+    /// Call <see cref="WriteAsync(IAsyncEnumerable{IngestionChunk{T}}, CancellationToken)"/> first.</exception>
+    public VectorStoreCollection<object, Dictionary<string, object?>> VectorStoreCollection
+        => _vectorStoreCollection ?? throw new InvalidOperationException("The collection has not been initialized yet. Call WriteAsync first.");
+
+    /// <inheritdoc/>
+    public override async Task WriteAsync(IAsyncEnumerable<IngestionChunk<T>> chunks, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(chunks);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        IReadOnlyList<object>? preExistingKeys = null;
+        await foreach (IngestionChunk<T> chunk in chunks.WithCancellation(cancellationToken))
+        {
+            if (_vectorStoreCollection is null)
+            {
+                _vectorStoreCollection = _vectorStore.GetDynamicCollection(_options.CollectionName, GetVectorStoreRecordDefinition(chunk));
+
+                await _vectorStoreCollection.EnsureCollectionExistsAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            // We obtain the IDs of the pre-existing chunks for given document,
+            // and delete them after we finish inserting the new chunks.
+            // To avoid a situation where we delete the chunks and then fail to insert the new ones.
```

Comment:
```suggestion
            // and delete them after we finish inserting the new chunks,
            // to avoid a situation where we delete the chunks and then fail to insert the new ones.
```